File: python/Binary_Search/day13/painterPartition.py
def countPainters(arr, max_units):
    painters = 1
    current_units = 0

    for num in arr:
        if current_units + num <= max_units:
            current_units += num
        else:
            painters += 1
            current_units = num

    return painters



def findUnits(arr, n, k):
    if k > n:
        return -1
    
    low = max(arr)
    high = sum(arr)


    # ---------- Linear Search ----------
    # for max_units in range(low, high + 1):
    #     requirePainters = countPainters(arr, max_units)

    #     if requirePainters <= k:
    #         return max_units

    # return -1





    # ---------- Binary Search ----------
    while low <= high:
        mid = (low + high) // 2

        requirePainters = countPainters(arr, mid)

        if requirePainters > k:
            low = mid + 1
        else:
            high = mid - 1
        
    return low







if __name__ == "__main__":
    n = int(input("Enter number of boards: "))
    arr = list(map(int, input("Enter units of each board: ").split()))
    k = int(input("Enter number of painters: "))

    ans = findUnits(arr, n, k)
    print("Minimum time required =", ans)

File: java/Binary_Search/Day13/PainterPartition.java
package Day13;
import java.util.*;

public class PainterPartition {
    public int countPainters(int[] arr, int units) {
        int painters = 1;
        long currentUnits = (long) 0;

        for (int num : arr) {
            if (currentUnits+ num <= units) {
                currentUnits += num;
            }
            else {
                painters++;
                currentUnits = num;
            }
        }
        return painters;
    }

    public int totalSum(int[] arr) {
        int total = 0;
        for (int num : arr) total += num;

        return total;
    }


    public int findUnits(int[] arr, int n, int k) {
        if (k > n) return -1;

        int low = Arrays.stream(arr).max().getAsInt();
        int high = totalSum(arr);

        // ---------- Linear Search ----------
        for (int units = low; units <= high; units++) {
            int requirePainters = countPainters(arr, units);

            if (requirePainters <= k) return units;
        }

        return -1;
    }





    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter number of boards: ");
        int n = sc.nextInt();

        int[] arr = new int[n];
        System.out.print("Enter units of each board: ");
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }

        System.out.print("Enter number of painters: ");
        int k = sc.nextInt();

        PainterPartition obj = new PainterPartition();
        int ans = obj.findUnits(arr, n, k);
        System.out.println("Minimum time required = " + ans);

        sc.close();
    }
}